edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter.Test/ConnectionEventTest.cs

```diff
@@ -0,0 +1,338 @@
+// Copyright (c) Microsoft. All rights reserved.namespace Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter.Test
+namespace Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter.Test
+{
+    using System;
+    using System.Collections.Generic;
+    using System.IO;
+    using System.Linq;
+    using System.Text;
+    using System.Text.RegularExpressions;
+    using System.Threading.Tasks;
+    using Microsoft.Azure.Devices.Edge.Hub.CloudProxy;
+    using Microsoft.Azure.Devices.Edge.Hub.Core;
+    using Microsoft.Azure.Devices.Edge.Hub.Core.Identity;
+    using Microsoft.Azure.Devices.Edge.Hub.Core.Routing;
+    using Microsoft.Azure.Devices.Edge.Hub.Core.Storage;
+    using Microsoft.Azure.Devices.Edge.Storage;
+    using Microsoft.Azure.Devices.Edge.Util;
+    using Microsoft.Azure.Devices.Edge.Util.Test.Common;
+    using Microsoft.Azure.Devices.Edge.Util.TransientFaultHandling;
+    using Microsoft.Azure.Devices.Routing.Core;
+    using Microsoft.Azure.Devices.Routing.Core.Endpoints;
+    using Microsoft.Azure.Devices.Shared;
+    using Newtonsoft.Json;
+    using Newtonsoft.Json.Bson;
+    using Newtonsoft.Json.Serialization;
+    using Moq;
+    using Xunit;
+
+    [Integration]
+    public class ConnectionEventTest
+    {
+        static readonly string iotHubName = "testHub";
+        static readonly string edgeModuleName = "$edgeHub";
+        static readonly string edgeDeviceId = "testDevice";
+
+        [Fact]
+        public async Task OnReconnectionClientsGetResubscribed()
+        {
+            // The logic of this test is the following
+            //  - First, generate the environment (edgeHub) with the necessary dependencies, so it can be emulated that
+            //    the MQTT broker sends connection-change-notifications, and on reconnect, edgeHub can execute its
+            //    resubscription cycle.
+            //  - The way we validate that edgeHub reacts to the connection notifications is that we catch the resubscriptions
+            //    on reconnect.
+            //  - The testcode generates clients (both module and device, and direct and nested), all clients subscribe to
+            //    desired property change, so after reconnection, edgeHub should re-subscribe for all clients.
+            var (subscriptionChangeHandler, cloudProxyDispatcher, brokerConnector) = await SetupEnvironment();
+            var allClients = await ConnectClientsAsync(subscriptionChangeHandler);
+
+            await Task.Delay(2000);
```

Comment:
How were these delay times chosen?

Reply:
Also, is it worthwhile making these values a constant somewhere in the test file if we are going to reuse the value?

Reply:
those are just 'long enough" time spans, so the async processing has time to finish its job. Moved into constant